docs/reference.md

```diff
@@ -131,3 +131,64 @@ dedicated test-suite, consisting of several tests exercising various runtime
 behaviors, for each StableHLO Op. The tests can be found
 [here](https://github.com/openxla/stablehlo/tree/main/stablehlo/tests/) (e.g.
 interpret\_\*.mlir).
+
+### Testing guidelines
+
+**(G1) Do we need to test for all the supported types for every op?**
+
+We can use a combination of following rules to decide it:
+
+  1. While implementing an op, if there exists code in the corresponding `eval`
+     function to handle a particular type, then it is imperative to have test(s)
+     to cover for that type. As an example, for `add` op, there is exclusive
+     code to handle integer, boolean, floating-point, and complex types, and
+     hence we need one test for each category of types.
+  2. If a set of types is handled uniformly in the corresponding
+     `eval` function, then a single test for all those types should be
+     sufficient. As an example, for `add` op, all the variants of integer
+     types (`si4`, `u4`, `si8`, `u8` and so on) are handled alike using
+     `llvm::APInt` APIs, and hence we can skip adding tests for each of those
+     variants, and instead, add a single representative test. To avoid ambiguity
+     in selecting the representative, we should use the following guidelines:
+
+       * If all the types, handled uniformaly, have the same primitive type
+         (i.e., if all are integer, or floating-point, or complex types), then
+         choose the one with maximum bit-width.
+       * If all the types, handled uniformaly, have a mix of primitive types,
+         then choose the one with the following primitive type, in decreasing
+         order of preference: integer, floating-point, boolean, complex.
```

Comment:
1. What is the rationale for this order?

Reply:
Thanks for bringing this up.

* If all the types, handled uniformaly, have a mix of primitive types,
         then choose..
An example of the ops which will fall into this bucket are ones in category "Data movement" like scatter/gather/sort etc.. 

One option I was considering  was to have the order same as https://github.com/openxla/stablehlo/blob/main/docs/spec.md#types.
which proposes to prefer boolean types over other types.
 
My current proposal is to order the types in a way such the the tests will add value to the understanding of the op.  

For example, writing tests, for say sort op, with boolean values vs integer values. IMO, the latter is better when it comes to understanding what the op is doing. This might be a trivial example, but the point is, in order to understand the semantics, often times we track the data movement to get an empirical understanding of the functioning of the op. For example, with gather/scatter, movements of 0/1 are much difficult to track that unique integer  values. 

Overall, the proposal is discouraging writing tests in boolean and complex over integers and floating-point values. 

The following (remaining) choices are  based on ease of representation. But I agree that it could be debatable.  
  - Between integer and floating-point, integer is preferred 
  - Between boolean and complex, boolean is preffered   
  
  Please let me know if this makes sense.

Reply:
Thanks for the detailed answer, between integer and floating-point, if floating point results do not provide any useful information from a precision standpoint then integer saves us a lot of hassle and gets the point across in terms of explaining the behaviour or logic of the op.